### PR TITLE
Product Instrumentation Pilot Tests

### PR DESCRIPTION
**Pull request by force2b**
Implements three pilot instrumentations for NPSP:

- When new Enhanced or Legacy Recurring Donations are created
- When new Engagement Plans are created
- When new Gifts are created through the Gift Entry UI - either as a single gift or as a batch of gifts

Work Item: [W-10398341](https://gus.lightning.force.com/lightning/r/a07EE00000dIModYAG/view)

# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: force-app/main/instrumentation/classes/SfdoInstrumentationMock_TEST.cls-meta.xml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>52.0</apiVersion>
```
Comment (bdvorachek) [inline force-app/main/instrumentation/classes/SfdoInstrumentationMock_TEST.cls-meta.xml:3]:
```suggestion
    <apiVersion>53.0</apiVersion>
```

File: force-app/main/default/classes/BDI_DataImportService.cls [hunk 2]
```diff
@@ -346,6 +346,36 @@ global with sharing class BDI_DataImportService {
         }
     }
 
+    /*******************************************************************************************************
+    * @description Instrument that multiple gift were successfully processed through the DataImport Batch process
+    * @param Id Apex Batch Job to find related DataImport__c records for
+    * @return void
+    */
+    public static void instrumentGiftEntryUIResults(Id apexJobId) {
```
Comment (bdvorachek) [inline force-app/main/default/classes/BDI_DataImportService.cls:354]: I think that either this function or the overload could get renamed. Does `instrumentGiftEntryBatchResults` or `instrumentGiftCount` feel like they'd work better?

File: force-app/main/default/classes/BDI_DataImportService.cls [hunk 3]
```diff
@@ -372,6 +402,11 @@ global with sharing class BDI_DataImportService {
     */
     global Boolean isDryRun { get; private set; }
 
+    /*******************************************************************************************************
+    * @description The source of the gift entry batch process - UI or someplace else
+    */
+    public static Boolean isFromGiftEntryUI = isExecutionSourceUserInterface();
```
Comment (bdvorachek) [inline force-app/main/default/classes/BDI_DataImportService.cls:408]: Can this drop down to:

```
@TestVisible
private static Boolean
```

I don't see it being set outside of a test class.

**Review by bdvorachek — CHANGES_REQUESTED** [review]

File: force-app/main/default/classes/BDI_DataImportService.cls (hunk 3 above)
Comment (force2b) [reply force-app/main/default/classes/BDI_DataImportService.cls]: It's used by the BDI_DataImport_BATCH class to get the state during instantiation.

**Review by bdvorachek — APPROVED** [review]

File: force-app/main/instrumentation/classes/SfdoInstrumentationMock_TEST.cls [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            if (stubbedMethodName == 'log') {
+                this.feature = (SfdoInstrumentationEnum.Feature) params[0];
+                this.component = (SfdoInstrumentationEnum.Component)params[1];
+                this.actionName = (SfdoInstrumentationEnum.Action)params[2];
+
+                for (Integer n=3; n<params.size(); n++) {
+                    if (paramNames[n] == 'value') {
+                        this.value =  (Integer)params[n];
+                    } else if (paramNames[n] == 'context') {
+                        this.context =  (Map<String,String>)params[n];
```
Comment (lparrott) [inline force-app/main/instrumentation/classes/SfdoInstrumentationMock_TEST.cls:76]: Does this still get used?

File: force-app/main/default/classes/BDI_DataImportService.cls [hunk 5]
```diff
@@ -346,6 +346,36 @@ global with sharing class BDI_DataImportService {
...
+    * @return void
+    */
+    public static void instrumentGiftEntry(Id apexJobId) {
+        try {
+            String createStatus = System.Label.bdiCreated;
+            Integer insertedGifts = [SELECT Count() 
+                FROM DataImport__c 
+                WHERE ApexJobId__c = :apexJobId
+                AND DonationImportStatus__c = :createstatus];
+                instrumentGiftEntry(insertedGifts);
```
Comment (lparrott) [inline force-app/main/default/classes/BDI_DataImportService.cls:361]: Bit of a nitpick, but can we un-tab this line? I thought it was part of the query at first glance!

**Review by lparrott — COMMENTED** [review]
Looks good, I just had a few comments/questions!

File: force-app/main/instrumentation/classes/SfdoInstrumentationMock_TEST.cls (hunk 4 above)
Comment (force2b) [reply force-app/main/instrumentation/classes/SfdoInstrumentationMock_TEST.cls:76]: This part of the stub (`context`) is not used yet, but I thought that keeping it in there would be helpful for when we do eventually start to instrument contextual content beyond a simple "value".

**Review by lparrott — APPROVED** [review]

**Comment by force2b** [comment]
Thank you both for the reviews. After QA is complete, we can consider merging this into the parent branch since it'll be the parent that ultimately needs to merge into main.

**Pull request merged**